src/framework/mlt_events.c

```diff
@@ -176,47 +174,33 @@ int mlt_events_register( mlt_properties self, const char *id, mlt_transmitter tr
 
 /** Fire an event.
  *
- * This takes a variable number of arguments to supply to the listener.
-
  * \public \memberof mlt_events_struct
  * \param self a properties list
  * \param id the name of an event
+ * \param data an opaque pointer
  * \return the number of listeners
  */
 
-int mlt_events_fire( mlt_properties self, const char *id, ... )
+int mlt_events_fire(mlt_properties self, const char *id, mlt_event_data *event_data)
```

Comment:
Suggestion: pass event_data by value

`int mlt_events_fire(mlt_properties self, const char *id, mlt_event_data event_data)`

Reply:
There are a lot of events with no data and currently pass NULL for this parameter. These would all have to be changed to pass basically a mlt_event_data somehow initialized with a NULL like
`mlt_event_data _from_object(NULL)`
or a new `mlt_event_data_null()`?

Does this change then motivate changing every `mlt_listener` to pass by value?
`typedef void ( *mlt_listener )( mlt_properties, void*, mlt_event_data );`
And then changing every all `mlt_event_data_to_...()` to pass by value? Finally, making all the changes everywhere these are used?

It is possible; I am in the middle of doing all that. It does make it easier and less verbose to use at the expense of more being more verbose when event data is null.

Reply:
I made all of these changes, and it is working. Overall I think it is cleaner, less pointer checks, Mlt::EventData no longer needs an `is_valid()`. Still reviewing all my changes before pushing it.

Reply:
Sounds promising.

For the "no data" cases, we could make a convenience function for that to set the value to 0:

```
mlt_event_data mlt_event_data_none()
{
    mlt_event_data event_data;
    event_data.u.i = 0;
    return event_data;
}
```

Depending on the patterns you see, you could also decide to specialize the fire function for each type:

```
mlt_events_fire(mlt_properties self, const char *id)
mlt_events_fire_data(mlt_properties self, const char *id, mlt_event_data *event_data)
mlt_events_fire_int(mlt_properties self, const char *id, int value)
mlt_events_fire_frame(mlt_properties self, const char *id, mlt_frame frame)
```

Based on your comments, it sounds like we may have already found the sweet spot.